File: trampoline.scala
import scala.annotation.tailrec


sealed trait Trampoline[+A]

object Trampoline {

    def call[A](expr: => Trampoline[A]): Call[A] = Call(() => expr)
    def done[A](value: A): Done[A] = Done(value)

}

case class Call[A](thunk: () => Trampoline[A]) extends Trampoline[A]
case class Done[A](value: A) extends Trampoline[A]


@annotation.tailrec
def trampoline[A](trmpln: Trampoline[A]): A = {
    trmpln match {
        case Call(thunk) => trampoline(thunk())
        case Done(value) => value
    }
}

def trampolineIsEven(n: Int): Trampoline[Boolean] = {
    if (n == 0) Trampoline.done(true)
    else Trampoline.call(trampolineIsOdd(n - 1))
}

def trampolineIsOdd(n: Int): Trampoline[Boolean] = {
    if (n == 0) Trampoline.done(false)
    else Trampoline.call(trampolineIsEven(n - 1))
}

def isEven(n: Int): Boolean = {
    trampoline(trampolineIsEven(n))
}

def isOdd(n: Int): Boolean = {
    trampoline(trampolineIsOdd(n))
}

File: trampoline.py
import functools


class Call(object):

    def __init__(self, func, *args, **kwargs):
        self.func = func
        self.args = args
        self.kwargs = kwargs

    def __call__(self):
        return self.func(*self.args, **self.kwargs)


class Done(object):
    def __init__(self, value):
        self.value = value

    def __call__(self):
        return self.value


def trampoline(func):

    @functools.wraps(func)
    def inner(*args, **kwargs):
        f = Call(func, *args, **kwargs)

        while isinstance(f, Call):
            f = f()
        return f()

    return inner


def trampoline_is_even(n):
    return Done(True) if n == 0 else Call(trampoline_is_odd, n - 1)

def trampoline_is_odd(n):
    return Done(False) if n == 0 else Call(trampoline_is_even, n - 1)


is_even = trampoline(trampoline_is_even)
is_odd = trampoline(trampoline_is_odd)
